futures-util/src/io/mod.rs

```diff
@@ -0,0 +1,150 @@
+                let b13: &mut [u8] = &mut [0];
+                let b14: &mut [u8] = &mut [0];
+                let b15: &mut [u8] = &mut [0];
+                let b16: &mut [u8] = &mut [0];
+                let mut bufs: [&mut IoVec; 16] = [
+                    b1.into(), b2.into(), b3.into(), b4.into(),
+                    b5.into(), b6.into(), b7.into(), b8.into(),
+                    b9.into(), b10.into(), b11.into(), b12.into(),
+                    b13.into(), b14.into(), b15.into(), b16.into(),
+                ];
+                let n = buf.bytes_vec_mut(&mut bufs);
+                try_ready!(self.poll_vectored_read(&mut bufs[..n], cx))
+            };
+
+            buf.advance_mut(n);
+            Ok(Async::Ready(n))
+        }
+    }
+
+    /// Provides a `Stream` and `Sink` interface for reading and writing to this
+    /// `Io` object, using `Decode` and `Encode` to read and write the raw data.
+    ///
+    /// Raw I/O objects work with byte sequences, but higher-level code usually
+    /// wants to batch these into meaningful chunks, called "frames". This
+    /// method layers framing on top of an I/O object, by using the `Codec`
+    /// traits to handle encoding and decoding of messages frames. Note that
+    /// the incoming and outgoing frame types may be distinct.
+    ///
+    /// This function returns a *single* object that is both `Stream` and
+    /// `Sink`; grouping this into a single object is often useful for layering
+    /// things like gzip or TLS, which require both read and write access to the
+    /// underlying object.
+    ///
+    /// If you want to work more directly with the streams and sink, consider
+    /// calling `split` on the `Framed` returned by this method, which will
+    /// break them into separate objects, allowing them to interact more easily.
+    fn framed<T: Encoder + Decoder>(self, codec: T) -> Framed<Self, T>
+        where Self: AsyncWrite + Sized,
+    {
+        framed::framed(self, codec)
+    }
+
+    /// Helper method for splitting this read/write object into two halves.
+    ///
+    /// The two halves returned implement the `Read` and `Write` traits,
+    /// respectively.
+    fn split(self) -> (ReadHalf<Self>, WriteHalf<Self>)
+        where Self: AsyncWrite + Sized,
+    {
+        split::split(self)
+    }
+}
+
+impl<T: AsyncRead + ?Sized> AsyncReadExt for T {}
+
+/// An extension trait which adds utility methods to `AsyncWrite` types.
+pub trait AsyncWriteExt: AsyncWrite {
+    /// Write a `Buf` into this value, returning how many bytes were written.
+    ///
+    /// Note that this method will advance the `buf` provided automatically by
+    /// the number of bytes written.
+    fn write_buf<B: Buf>(&mut self, buf: &mut B, cx: &mut task::Context)
+        -> Poll<usize, std_io::Error>
+        where Self: Sized,
+    {
+        if !buf.has_remaining() {
+            return Ok(Async::Ready(0));
+        }
+
+        let n = try_ready!(self.poll_write(buf.bytes(), cx));
```

Comment:
Seems the default implementation here could make use of `self.poll_vectored_write`.

Reply:
Done! Thanks for the reminder :)